src/fold.rs

```diff
@@ -24,17 +25,27 @@ macro_rules! fold_int_bin_op {
     }
 }
 
-macro_rules! fold_scalar_bin_op {
-    ($op: tt) => {
-        |a: &Literal, b: &Literal, _| match (a, b) {
-            (Int(a), Int(b)) => Ok(Some(Int(a $op b))),
-            (UnsignedInt(a), UnsignedInt(b)) => Ok(Some(UnsignedInt(a $op b))),
-            (Float(a), Float(b)) => Ok(Some(Float(a $op b))),
-            (Char(a), Char(b)) => Ok(Some(Char(a $op b))),
-            // TODO: find a way to do this that allows `"hello" + 2 - 1`
-            //(Str(s), Int(i)) | (Int(i), Str(s)) => {
-            (_, _) => Ok(None),
+#[inline]
+fn fold_scalar_bin_op(
+    simple: fn(f64, f64) -> f64,
+    overflowing: fn(i64, i64) -> (i64, bool),
+    wrapping: fn(u64, u64) -> u64,
+) -> impl Fn(&Literal, &Literal, &Type) -> Result<Option<Literal>, Error> {
+    move |a: &Literal, b: &Literal, _ctype| match (a, b) {
+        (Int(a), Int(b)) => {
+            let (value, overflowed) = overflowing(*a, *b);
```

Comment:
Can you comment here that `overflowing` returns the wrapped value? I was confused the first time reading it.
```suggestion
            // `overflowing` returns the _wrapped_ value,
            //  so if we had negative overflow `value` would be positive and vice-versa.
            let (value, overflowed) = overflowing(*a, *b);
```